### PR TITLE
Fixed race condition in the state reducers. Also fixed missing reducer

### PR DESCRIPTION
**Pull request by hellohuanlin**
Fix the race condition introduced here https://github.com/airbnb/MvRx/pull/206
Also added detailed explanation on the job scheduling algorithm

1. Fixed the race condition

let's say we have `[G1, G2, G3]` in getStateQueue and `[S1, S2, S3]` in setStateQueue. we want to purge all [S1, S2, S3] before taking any task in getStateQueue. With the original implementation, the execution order will be `S1 -> G1 -> S2 -> G2 -> S3 -> G3`, which has the race condition. With this change, the execution order will be `S1 -> S2 -> S3 -> G1 -> G2 -> G3`

2. Also fix a missing reducer. 

Let's say we have [G1] and [S1, S2, S3], in the original implementation, execution order will be `S1 -> G1 -> S2` (`S3` is missing). with this PR update, execution order will be `S1 -> S2 -> S3 -> G1`

@gpeal @BenSchwab @elihart

### CODE REVIEW

File: mvrx/src/main/kotlin/com/airbnb/mvrx/RealMvRxStateStore.kt [hunk 1]
```diff
@@ -120,10 +187,13 @@ class RealMvRxStateStore<S : Any>(initialState: S) : MvRxStateStore<S> {
     }
 
     private fun flushSetStateQueue() {
```
Comment (BenSchwab) [inline mvrx/src/main/kotlin/com/airbnb/mvrx/RealMvRxStateStore.kt]: @hellohuanlin isn't this called `flushQueues()` which does drain the setState queue. Not sure what is changing here. Could you clarify?

File: mvrx/src/main/kotlin/com/airbnb/mvrx/RealMvRxStateStore.kt [hunk 2]
```diff
@@ -78,6 +78,67 @@ class RealMvRxStateStore<S : Any>(initialState: S) : MvRxStateStore<S> {
         flushQueueSubject.onNext(Unit)
     }
 
+
+    /**
```
Comment (BenSchwab) [inline mvrx/src/main/kotlin/com/airbnb/mvrx/RealMvRxStateStore.kt]: Thanks for adding this comment!

File: mvrx/src/main/kotlin/com/airbnb/mvrx/RealMvRxStateStore.kt (hunk 1 above)
Comment (hellohuanlin) [reply mvrx/src/main/kotlin/com/airbnb/mvrx/RealMvRxStateStore.kt]: let's say we have `[G1, G2, G3]` in `getStateQueue` and `[S1, S2, S3]` in `setStateQueue`. we want to purge all `[S1, S2, S3]` before taking any task in `getStateQueue`. With the original implementation, the execution order will be `S3 -> G3 -> S2 -> G2 -> S1 -> G1`, which has the race condition. With this change, the execution order will be `S3 -> S2 -> S1 -> G3 -> G2 -> G1`

Comment (hellohuanlin) [reply mvrx/src/main/kotlin/com/airbnb/mvrx/RealMvRxStateStore.kt]: Oh, actually if you follow the execution path, the original implementation actually may have missing reducers. 

Let's say we have `[G1]` and `[S1, S2, S3]`, in the original implementation, execution order will be `S1 -> G1 -> S2` (`S3` is missing)

So this PR not just fix the race condition, but also fix the missing reducer.

Comment (hellohuanlin) [reply mvrx/src/main/kotlin/com/airbnb/mvrx/RealMvRxStateStore.kt]: I will update the PR description

**Comment by hellohuanlin** [comment]
@gpeal bump

**Comment by hellohuanlin** [comment]
@gpeal yeah i figured that it was probably introduced by mistake. Sorry I didn't see your comment earlier. Will add more tests for this in case of future regression.

**Pull request merged**